### ISSUE TITLE
Simultanously operating on multiple atomspaces within guile

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1393]
I'm having a problem with making the following multithreaded because I can have 2 processes in guile work on their own atomspace. Here's how my experiment works

1. Run N problems on N different atomspaces and generate a history atomspace per problem

2. Merge the history atomspaces into one and Perform learning on the global history

3. Repeat 1

1 is computationally hungry and is fortunately embarrassingly parallelizable IF each process can work on its own atomspace.

Since the pattern matcher (and ultimately the URE) scheme bindings don't allow the user to call them on another atomspace than the current one, I can't parallelize 1.

The only things I can think of to address that are

1. change the pattern matcher and URE scheme binding to accept another atomspace than the current one

or

2. Move each process computation in a single guile process and have bash (or guile itself) system call them, etc.

Any suggestion?

**Comment by linas** [comment]
Do both.

1.  There is a way in scheme to add optional arguments, and you could do that.  But please don't make changes that require 100 different files to be touched.  That's very disruptive.  An easier, better thing to do would be to create all-new bindings that take the atomspace as an argument, play with them, test them, and 6 months later, when they work well, port the old scheme bindings to sit on top of the new ones.  

1.a. Lots of threads are hard to manage and debug.  If/when something crashes, then full-stop.   If there is memory corruption, its hard to debug. The guile GC does struggle in certain large-RAM situations.  The built-in guile threads totally suck; the built-in parallel-for-each in guile is a total fail for more than 3 threads.  I don't know why, its sad.  For these reasons, option 2 is nice...

2. We need a much better multi-server system/infrastructure.  So, for example, right now we have "gearman", which allows you to start a bunch of cogservers, even on different machines, and send them scheme strings, to run something, and then send back scheme strings to the central cogserver. 

Mandeep wrote the gearman module. We have a unit test for it. It worked when I tried it. Its probably not documented. Its maybe hard to configure? I've never used it for anything more complicated than the unit test.  This might be a good time for it to be dusted off and improved.

(I don't think it needs to be replaced. At least, not yet. I like it because it is very simple, very easy, and I would like to keep things simply and easy, rather than over-engineering them at the start.  We can do more engineering later on, once we understand multi-atomspace algos better.)

**Comment by ngeiswei** [comment]
Thanks Linas! I won't bother with 1 then. I might go with 2, even if that requires work because it seems very valuable in the long term.

That's all I could find about gearman and opencog

http://wiki.opencog.org/w/API#Distributed_computing
https://github.com/opencog/atomspace/tree/master/opencog/persist/gearman
http://wiki.opencog.org/w/Suggested_Projects#Write_a_test_application_for_the_Gearman_distributed-querying_infrastructure

Let me know if I missed something.

**Comment by ngeiswei** [comment]
I'll add a third option which is to set a persistent storage, and use gnu parallel https://www.gnu.org/software/parallel/ to dispatch the work.

**Comment by linas** [comment]
I'm sorry that gearman is turning out to be buggy.... the dispatch to
workers should be non blocking.

**Comment by ngeiswei** [comment]
I'm gonna try to use a persistent storage, is the page http://wiki.opencog.org/w/Using_a_Backing_Store up to date?

**Comment by AmeBel** [comment]
https://github.com/opencog/atomspace/tree/master/opencog/persist/sql#install-setup-and-usage-howto is likely to be more up to date.

**Comment by linas** [comment]
The wiki page is stale; the ODBC drivers are deprecated.  The postgres drivers are faster and much easier to configure.

Using postgres to communicate between multiple servers should work; however, there is no built-in sync primitives to guarantee that the sender has finished sending before the receiver can assume the data is ready.  There is a fence/flush primitive in the opencog api, but it does NOT do a round-trip to the database.

**Comment by ngeiswei** [comment]
@linas in https://github.com/opencog/atomspace/blob/master/opencog/persist/sql/README.md#missing-featurestodo-items it says "Add support for multiple atom spaces.". Does that mean you can't have more one atomspace in persistent storage?

**Comment by ngeiswei** [comment]
@linas you said

"There is a fence/flush primitive in the opencog api, but it does NOT do a round-trip to the database" 

is it still true if `synchronous_commit=local`?

**Comment by linas** [comment]
>  Does that mean you can't have more one atomspace in persistent storage?

Yes.  The database tables have a column reserved for an atomspace uuid number, but it is not used.  The primary design issue is knowing which atomspace is which. Clearly, "atomspace 1" can be the primary atomspace, but if there are more, how does a client know what they mean, what they contain?  Should atomspaces be given a name? (as opposed to some random UUID identifying the atomspace).  Should there be an `AtomSpaceNode` to hold that name?  Or maybe an `AtomSpaceLink`?  How could this work?

**Comment by linas** [comment]
>>  "There is a fence/flush primitive in the opencog api, but it does NOT do a round-trip to the database"

> is it still true if `synchronous_commit=local?`

Yes, I think so. Flushing is complicated: The atoms may be stuck in the opencog write-back queue, and not yet given to the postgres client library. Even if given to the postgres client library, they may be stuck in the operating-system tcp/unix socket, so that the postgres server hasn't yet seen them. 

I believe that `synchronous_commit` is all about writing to disk, but that is not what matters for multi-client communication.  What does matter is that if client A sends some atoms to the postgres server, the postgres server will see them, so that when client B asks for them, the postgres server can respond. It really doesn't matter if they've been written to disk or not; it does matter that they are not stuck in some tcp socket or buffer. 

In general, you don't want to do this kind of multi-client synchronization for *every* atom; that would be expensive. Instead, you want to do it only when its actually needed. 

Implementing this kind of synchronization is not "hard", but it does require reading train-loads of blogs and opinions and comment sections on postgres discussion sites, as otherwise, you'll miss some important issue.

There might be some opencog issues. If client A and client B both create `ConceptNode "foo"` and both try to send it to the database for the first time, at the same time, will there be a collision? Error? there should not be but who knows ... The uuid needs to be consistently assigned across both clients, its not clear that this is done right now.  It might be, not sure, the code needs review.

**Comment by ngeiswei** [comment]
@linas you said

"There is a fence/flush primitive in the opencog api, but it does NOT do a round-trip to the database"

If a process close the database, via `(sql-close)`, after storing atoms, doesn't that guaranty that these atoms are available to the other processes? If not, should it?

**Comment by linas** [comment]
> `(sql-close)`

No. There is no  synchronizing primitive in the API right now. 

If you call `(store-atom foo)`, the atom `foo` will **eventually** be written to the database, and that is guaranteed. It just probably won't be written before the call returns. 

The `(sql-close)` statement does not offer any additional guarantees.  It just closes the connection; that is all.  It may return before any atoms are written to the database.  It is not a synchronizing primitive.  It should not be, roughly because synchronization is unrelated to connection handling, both conceptually, and in practice.  The code needed to force synchronization has nothing to do with managing the connection.
